### DISCUSSION TITLE
Loading ActiveRelation Model with dataloader but want to paginate it

### DISCUSSION THREAD

**Discussion by tomocrafter** [rmosolgo/graphql-ruby#4817] [Q&A] [answered]
Hi, I am currently tring to load model with built in dataloader and paginate it with connection_type.
There are almost 100 ids in a config file and field that returns by pagination. I wrote the resolver like following:

```ruby
field :products, Types::Objects:Product.connection_type, null: false
def products
  dataloader.with(Sources::ProductById).load_all(prepared_ids)
end
```

but it loads all product that in `prepared_ids`. so the connection type will be ArrayConnection.
How can I load only required id products which applied connection arguments (like cursor, offset, limit) ?

Thank you.

**Comment by rmosolgo** [comment, marked as answer]
Hi! I think your best bet is to return an ActiveRecord relation from the resolver, for example:

```ruby 
def products 
  Product.where(id: prepared_ids)
end 
```

Then, GraphQL-Ruby's `ActiveRecordRelationConnection` will apply limit and offset according the client's provided arguments. 

In short, batch-loading relations is very hard because you can't load different pages of different relations at the same time. AFAIK only Postgresql "Window functions" can do this, see https://pganalyze.com/blog/efficient-graphql-queries-in-ruby-on-rails-and-postgres#batch-loading-many-records for a discussion of why. I hope that helps!

**Reply by tomocrafter** [reply]
Thank you for your response.
It appears that batching relations is indeed challenging. I have encountered a similar issue before.
Currently, I am using PostgreSQL for the database, so I am open to methods that utilize window functions. However, when attempting to apply this to a connection_type, I wonder if I need to manually control the behavior of limits and cursors. Is there a simpler way to implement this by using features like ArrayConnection or ConnectionBehavior?
